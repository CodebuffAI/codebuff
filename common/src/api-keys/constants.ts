export const ALGORITHM = 'aes-256-gcm'
export const IV_LENGTH = 12
export const AUTH_TAG_LENGTH = 16

// --- Constants for API Key Validation ---
export const KEY_PREFIXES = {
  anthropic: 'sk-ant-api03-',
  gemini: 'AIzaSy',
  openai: 'sk-proj-',
}
export const KEY_LENGTHS = {
  anthropic: 108,
  gemini: 39,
  openai: 164,
}
