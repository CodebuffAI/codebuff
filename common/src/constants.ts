export const STOP_MARKER = '[' + 'END]'
export const FIND_FILES_MARKER = '[' + 'FIND_FILES_PLEASE]'
export const EXISTING_CODE_MARKER = '[[**REPLACE_WITH_EXISTING_CODE**]]'

export const DEFAULT_IGNORED_FILES = [
  '.git',
  '.env',
  '.env.*',
  'env',
  'ENV',
  '*.min.*',
  'node_modules',
  'venv',
  'virtualenv',
  '.venv',
  '.virtualenv',
  '__pycache__',
  '*.egg-info/',
  '*.pyc',
  '.DS_Store',
  '.pytest_cache',
  '.mypy_cache',
  '.ruff_cache',
  '.next',
  'package-lock.json',
  'bun.lockb',
]

// Special message content tags indicating specific server states
export const ASKED_CONFIG = 'asked_config'
export const SHOULD_ASK_CONFIG = 'should_ask_config'
export const ONE_TIME_TAGS = [] as const
export const ONE_TIME_LABELS = [
  ...ONE_TIME_TAGS,
  ASKED_CONFIG,
  SHOULD_ASK_CONFIG,
] as const

export const FILE_READ_STATUS = {
  DOES_NOT_EXIST: '[FILE_DOES_NOT_EXIST]',
  IGNORED: '[FILE_IGNORED_BY_GITIGNORE_OR_CODEBUFF_IGNORE]',
  OUTSIDE_PROJECT: '[FILE_OUTSIDE_PROJECT]',
  TOO_LARGE: '[FILE_TOO_LARGE]',
  ERROR: '[FILE_READ_ERROR]',
} as const

export const HIDDEN_FILE_READ_STATUS = [
  FILE_READ_STATUS.DOES_NOT_EXIST,
  FILE_READ_STATUS.IGNORED,
  FILE_READ_STATUS.OUTSIDE_PROJECT,
  FILE_READ_STATUS.TOO_LARGE,
  FILE_READ_STATUS.ERROR,
]

export function toOptionalFile(file: string | null) {
  if (file === null) return null
  return HIDDEN_FILE_READ_STATUS.some((status) => file.startsWith(status))
    ? null
    : file
}

export const REQUEST_CREDIT_SHOW_THRESHOLD = 1
export const MAX_DATE = new Date(86399999999999)
export const BILLING_PERIOD_DAYS = 30
export const CREDITS_REFERRAL_BONUS = 250
export const AFFILIATE_USER_REFFERAL_LIMIT = 500

// Default number of free credits granted per cycle
export const DEFAULT_FREE_CREDITS_GRANT = 500

export const AuthState = {
  LOGGED_OUT: 'LOGGED_OUT',
  LOGGED_IN: 'LOGGED_IN',
} as const

export type AuthState = (typeof AuthState)[keyof typeof AuthState]

export const UserState = {
  LOGGED_OUT: 'LOGGED_OUT',
  GOOD_STANDING: 'GOOD_STANDING', // >= 100 credits
  ATTENTION_NEEDED: 'ATTENTION_NEEDED', // 20-99 credits
  CRITICAL: 'CRITICAL', // 1-19 credits
  DEPLETED: 'DEPLETED', // <= 0 credits
} as const

export type UserState = (typeof UserState)[keyof typeof UserState]

export function getUserState(isLoggedIn: boolean, credits: number): UserState {
  if (!isLoggedIn) return UserState.LOGGED_OUT

  if (credits >= 100) return UserState.GOOD_STANDING
  if (credits >= 20) return UserState.ATTENTION_NEEDED
  if (credits >= 1) return UserState.CRITICAL
  return UserState.DEPLETED
}

export const costModes = ['lite', 'normal', 'max', 'experimental'] as const
export type CostMode = (typeof costModes)[number]

export const getModelForMode = (
  costMode: CostMode,
  operation: 'agent' | 'file-requests' | 'check-new-files'
) => {
  if (operation === 'agent') {
    return {
      lite: models.gemini2_5_flash_thinking,
      normal: models.sonnet,
      max: models.sonnet,
      experimental: models.gemini2_5_pro_preview,
    }[costMode]
  }
  if (operation === 'file-requests') {
    return {
      lite: claudeModels.haiku,
      normal: claudeModels.haiku,
      max: claudeModels.sonnet,
      experimental: claudeModels.sonnet,
    }[costMode]
  }
  if (operation === 'check-new-files') {
    return {
      lite: models.gpt4omini,
      normal: models.gpt4o,
      max: models.gpt4o,
      experimental: models.gpt4o,
    }[costMode]
  }
  throw new Error(`Unknown operation: ${operation}`)
}

export const claudeModels = {
  sonnet: 'claude-3-5-sonnet-20241022',
  sonnet3_7: 'claude-3-7-sonnet-20250219',
  haiku: 'claude-3-5-haiku-20241022',
} as const
export type AnthropicModel = (typeof claudeModels)[keyof typeof claudeModels]

export const openaiModels = {
  gpt4_1: 'gpt-4.1-2025-04-14',
  gpt4o: 'gpt-4o-2024-11-20',
  gpt4omini: 'gpt-4o-mini-2024-07-18',
  o3mini: 'o3-mini-2025-01-31',
  o3: 'o3-2025-04-16',
  o4mini: 'o4-mini-2025-04-16',
  generatePatch:
    'ft:gpt-4o-2024-08-06:manifold-markets:generate-patch-batch2:AKYtDIhk',
} as const
export type OpenAIModel = (typeof openaiModels)[keyof typeof openaiModels]

export const geminiModels = {
  gemini2_5_flash: 'gemini-2.5-flash-preview-04-17',
  gemini2_5_flash_thinking: 'gemini-2.5-flash-preview-04-17:thinking',
  gemini2flash: 'gemini-2.0-flash-001',
  gemini2_5_pro_exp: 'gemini-2.5-pro-exp-03-25',
  gemini2_5_pro_preview: 'gemini-2.5-pro-preview-05-06',
} as const
export type GeminiModel = (typeof geminiModels)[keyof typeof geminiModels]

export const openrouterModels = {
  openrouter_gemini2_5_pro_exp: 'google/gemini-2.5-pro-exp-03-25:free',
  openrouter_gemini2_5_pro_preview: 'google/gemini-2.5-pro-preview-03-25',
  openrouter_gemini2_5_flash: 'google/gemini-2.5-flash-preview',
  openrouter_gemini2_5_flash_thinking:
    'google/gemini-2.5-flash-preview:thinking',
} as const
export type openrouterModel =
  (typeof openrouterModels)[keyof typeof openrouterModels]

export const deepseekModels = {
  deepseekChat: 'deepseek-chat',
  deepseekReasoner: 'deepseek-reasoner',
} as const
export type DeepseekModel = (typeof deepseekModels)[keyof typeof deepseekModels]

// Vertex uses "endpoint IDs" for finetuned models, which are just integers
export const finetunedVertexModels = {
  ft_filepicker_003: '196166068534771712',
  ft_filepicker_005: '8493203957034778624',
} as const
export type FinetunedVertexModel =
  (typeof finetunedVertexModels)[keyof typeof finetunedVertexModels]

export const models = {
  ...claudeModels,
  ...openaiModels,
  ...geminiModels,
  ...deepseekModels,
  ...openrouterModels,
  ...finetunedVertexModels,
} as const

export const shortModelNames = {
  'gemini-2.5-pro': models.gemini2_5_pro_preview,
  'sonnet-3.7': models.sonnet3_7,
  'sonnet-3.5': models.sonnet,
  'sonnet-3.6': models.sonnet,
  'gpt-4.1': models.gpt4_1,
  'o3-mini': models.o3mini,
  o3: models.o3,
  'o4-mini': models.o4mini,
}

export const providerModelNames = {
  [models.gemini2_5_pro_preview]: 'gemini',
  [models.gemini2_5_pro_exp]: 'gemini',
  [models.gemini2flash]: 'gemini',
  [models.gemini2_5_flash]: 'gemini',
  [models.gemini2_5_flash_thinking]: 'gemini',
  [models.haiku]: 'anthropic',
  [models.sonnet3_7]: 'anthropic',
  [models.sonnet]: 'anthropic',
  [models.gpt4_1]: 'openai',
  [models.gpt4o]: 'openai',
  [models.gpt4omini]: 'openai',
  [models.o3mini]: 'openai',
  [models.o3]: 'openai',
  [models.o4mini]: 'openai',
}

export type Model = (typeof models)[keyof typeof models]

export const TEST_USER_ID = 'test-user-id'
