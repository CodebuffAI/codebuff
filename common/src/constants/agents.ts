// Define agent personas with their shared characteristics
export const AGENT_PERSONAS = {
  // Base agents - all use Buffy persona
  base: {
    displayName: 'Buffy the Enthusiastic Coding Assistant',
    purpose: 'Base agent that orchestrates the full response.',
  },
  base_lite: {
    displayName: 'Buffy the Enthusiastic Coding Assistant',
    purpose: 'Base agent that orchestrates the full response.',
  },
  base_max: {
    displayName: 'Buffy the Enthusiastic Coding Assistant',
    purpose: 'Base agent that orchestrates the full response.',
  },
  base_experimental: {
    displayName: 'Buffy the Enthusiastic Coding Assistant',
    purpose: 'Base agent that orchestrates the full response.',
  },
  claude4_gemini_thinking: {
    displayName: 'Buffy the Enthusiastic Coding Assistant',
    purpose: 'Base agent that orchestrates the full response.',
  },

  superagent: {
    displayName: 'Superagent',
    purpose:
      'Superagent that can spawn multiple code editing agents to complete a task.',
  },

  // Ask mode
  ask: {
    displayName: 'Buffy the Enthusiastic Coding Assistant',
    purpose: 'Base ask-mode agent that orchestrates the full response.',
  },

  // Specialized agents
  thinker: {
    displayName: 'Theo the Theorizer',
    purpose:
      'Does deep thinking given the current messages and a specific prompt to focus on. Use this to help you solve a specific problem.',
  },
  file_explorer: {
    displayName: 'Dora The File Explorer',
    purpose: 'Expert at exploring a codebase and finding relevant files.',
  },
  file_picker: {
    displayName: 'Fletcher the File Fetcher',
    purpose: 'Expert at finding relevant files in a codebase.',
  },
  researcher: {
    displayName: 'Reid Searcher the Researcher',
    purpose: 'Expert at researching topics using web search and documentation.',
  },
  planner: {
    displayName: 'Peter Plan the Planner',
    purpose: 'Agent that formulates a comprehensive plan to a prompt.',
    hidden: true,
  },
  dry_run: {
    displayName: 'Sketch the Dry Runner',
    purpose: 'Agent that takes a plan and try to implement it in a dry run.',
    hidden: true,
  },
  reviewer: {
    displayName: 'Nit Pick Nick the Reviewer',
    purpose:
      'Reviews file changes and responds with critical feedback. Use this after making any significant change to the codebase.',
  },
  sonnet4_agent_builder: {
    displayName: 'Agna the Agent Builder',
    purpose: 'Creates new agent templates for the codebuff mult-agent system',
    hidden: true,
  },
} as const

// Agent IDs list from AGENT_PERSONAS keys
export const AGENT_IDS = Object.keys(
  AGENT_PERSONAS
) as (keyof typeof AGENT_PERSONAS)[]

// Agent ID prefix constant
export const AGENT_ID_PREFIX = 'CodebuffAI/'

// Agent names for client-side reference
export const AGENT_NAMES = Object.fromEntries(
  Object.entries(AGENT_PERSONAS).map(([agentType, persona]) => [
    agentType,
    persona.displayName,
  ])
) as Record<keyof typeof AGENT_PERSONAS, string>

export type AgentName =
  (typeof AGENT_PERSONAS)[keyof typeof AGENT_PERSONAS]['displayName']

// Get unique agent names for UI display
export const UNIQUE_AGENT_NAMES = Array.from(
  new Set(
    Object.values(AGENT_PERSONAS)
      .filter((persona) => !(persona as any).hidden)
      .map((persona) => persona.displayName)
  )
)

// Map from display name back to agent types (for parsing user input)
export const AGENT_NAME_TO_TYPES = Object.entries(AGENT_NAMES).reduce(
  (acc, [type, name]) => {
    if (!acc[name]) acc[name] = []
    acc[name].push(type)
    return acc
  },
  {} as Record<string, string[]>
)
