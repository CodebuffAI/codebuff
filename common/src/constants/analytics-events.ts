/**
 * Enum of analytics event types used throughout the application
 */
export enum AnalyticsEvent {
  // CLI
  APP_LAUNCHED = 'cli.app_launched',
  BACKGROUND_PROCESS_CONTINUE = 'cli.background_process_continue',
  BACKGROUND_PROCESS_END = 'cli.background_process_end',
  BACKGROUND_PROCESS_LEFTOVER_DETECTED = 'cli.background_process_leftover_detected',
  BACKGROUND_PROCESS_START = 'cli.background_process_start',
  CHANGE_DIRECTORY = 'cli.change_directory',
  CHECKPOINT_COMMAND_USED = 'cli.checkpoint_command_used',
  LOGIN = 'cli.login',
  TERMINAL_COMMAND_COMPLETED = 'cli.terminal_command_completed',
  MALFORMED_PROMPT_RESPONSE = 'cli.malformed_prompt_response',

  // Backend
  USER_INPUT = 'backend.user_input',
  AGENT_STEP = 'backend.agent_step',
  CREDIT_GRANT = 'backend.credit_grant',
  CREDIT_CONSUMED = 'backend.credit_consumed',
  TOOL_USE = 'backend.tool_use',

  // Web
  SIGNUP = 'web.signup',

  // Common
  FLUSH_FAILED = 'common.flush_failed',
}
