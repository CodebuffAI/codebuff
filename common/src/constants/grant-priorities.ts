import type { GrantType } from '../db/schema'

export const GRANT_PRIORITIES: Record<GrantType, number> = {
  free: 20,
  referral: 40,
  admin: 60,
  organization: 70,
  purchase: 80,
} as const
