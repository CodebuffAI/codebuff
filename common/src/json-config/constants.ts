import { z } from 'zod'

export const codebuffConfigFile = 'codebuff.json'
export const codebuffConfigFileBackup = 'codebuff.jsonc'

export const StartupProcessSchema = z
  .object({
    name: z
      .string()
      .min(1, 'Process name is required')
      .describe(
        'A user-friendly name for the process. Should be one word and unique.'
      ),
    command: z
      .string()
      .min(1, 'Command is required')
      .describe('The actual shell command to execute.'),
    cwd: z
      .string()
      .optional()
      .describe('The working directory from which to run the command.'),
    enabled: z
      .boolean()
      .optional()
      .default(true)
      .describe('Whether this process should be run'),
    stdoutFile: z
      .string()
      .optional()
      .describe(
        "Path to write the process's stdout. If not specified, stderr is not stored."
      ),
    stderrFile: z
      .string()
      .optional()
      .describe(
        "Path to write the process's stderr. If not specified, stderr will be put into the stdoutFile."
      ),
  })
  .describe('Defines a single startup process.')

export const CodebuffConfigSchema = z
  .object({
    description: z
      .any()
      .optional()
      .describe('Does nothing. Put any thing you want here!'),
    startupProcesses: z
      .array(StartupProcessSchema)
      .optional()
      .describe('An array of startup processes.'),
    maxAgentSteps: z
      .number()
      .default(12)
      .describe(
        'Maximum number of turns agent will take before being forced to end'
      ),
  })
  .describe(
    `Defines the overall Codebuff configuration file (e.g., ${codebuffConfigFile}). This schema defines the top-level structure of the configuration.`
  )

/**
 * TypeScript type representing a validated startup process object.
 * This type is inferred from the `StartupProcessSchema` and provides type safety
 * when working with startup process configurations in code.
 */
export type StartupProcess = z.infer<typeof StartupProcessSchema>

/**
 * TypeScript type representing a validated Codebuff configuration object.
 * This type is inferred from the `CodebuffConfigSchema` and provides type safety
 * for the entire configuration structure.
 */
export type CodebuffConfig = z.infer<typeof CodebuffConfigSchema>
