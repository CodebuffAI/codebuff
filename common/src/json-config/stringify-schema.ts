import { z, type ZodTypeAny } from 'zod'

function getSchemaDescription(schema: ZodTypeAny): string | undefined {
  return schema.description
}

function indent(level: number): string {
  return ' '.repeat(level * 2)
}

function isOptionalType(schema: ZodTypeAny): boolean {
  if (schema instanceof z.ZodOptional) return true
  if (schema instanceof z.ZodNullable && schema._def.innerType instanceof z.ZodOptional) return true
  if (schema instanceof z.ZodDefault && schema._def.innerType instanceof z.ZodOptional) return true
  return false
}

function stringifyZodType(
  schema: ZodTypeAny,
  level: number = 0,
  fieldName?: string,
  needsComma: boolean = false,
  includeDescription: boolean = true
): string {
  const description = includeDescription ? getSchemaDescription(schema) : undefined
  const baseIndent = indent(level)
  const fieldPrefix = fieldName ? `"${fieldName}": ` : ''
  const comma = needsComma ? ',' : ''

  // Handle object types
  if (schema instanceof z.ZodObject) {
    const shape = schema.shape
    let output = ''
    if (description) {
      output += `${baseIndent}// ${description}\n`
    }
    output += `${baseIndent}${fieldPrefix}{\n`
    
    if (Object.keys(shape).length === 0) {
      return output + baseIndent + '}'
    }

    const entries = Object.entries(shape)
    
    entries.forEach(([key, value], index) => {
      const isLastLine = index === entries.length - 1
      const valueDescription = getSchemaDescription(value as ZodTypeAny)
      const isOptional = value instanceof z.ZodOptional || value instanceof z.ZodDefault
      const defaultValue = value instanceof z.ZodDefault ? JSON.stringify(value._def.defaultValue()) : undefined
      
      if (valueDescription) {
        const prefix = isOptional ? '(optional): ' : ''
        const suffix = defaultValue ? `, default: ${defaultValue}` : ''
        // Add newline before comment, but not for first item
        if (index > 0) output += '\n'
        output += `${indent(level + 1)}// ${prefix}${valueDescription}${suffix}\n`
      }
      
      const typeStr = stringifyZodType(value as ZodTypeAny, 0, undefined, false, false)
      output += `${indent(level + 1)}"${key}": ${typeStr}${isLastLine ? '' : ','}\n`
    })

    output += baseIndent + '}'
    return output + (needsComma ? ',' : '')
  }

  // Handle array types
  if (schema instanceof z.ZodArray) {
    let output = ''
    if (description) {
      output += `${baseIndent}// ${description}\n`
    }
    output += `${baseIndent}${fieldPrefix}[\n\n`
    const elementType = schema.element
    if (elementType instanceof z.ZodObject) {
      // For objects, we want the opening brace to be indented two levels
      const objectOutput = stringifyZodType(elementType, level + 2, undefined, false, true)
      output += objectOutput
    } else {
      // For non-objects, we just want the type name
      output += indent(level + 1) + stringifyZodType(elementType, 0, undefined, false, false)
        .trim()
        .replace(/^".*":\s*/, '') // Remove field name if present
    }
    output += `\n${indent(level + 1)}]${comma}`
    return output
  }

  // Handle optional types
  if (schema instanceof z.ZodOptional || schema instanceof z.ZodNullable) {
    const innerType = stringifyZodType(schema._def.innerType, level, undefined, false, false)
      .trim()
      .replace(/^".*":\s*/, '') // Remove field name if present
    const nullValue = schema instanceof z.ZodNullable ? 'null' : 'undefined'
    let output = ''
    if (description && includeDescription) {
      output += `${baseIndent}// ${description}\n${baseIndent}`
    }
    return output + `${fieldPrefix}${innerType} | ${nullValue}${comma}`
  }

  // Handle default types
  if (schema instanceof z.ZodDefault) {
    const innerType = stringifyZodType(schema._def.innerType, level, undefined, false, false)
      .trim()
      .replace(/^".*":\s*/, '') // Remove field name if present
    let output = ''
    if (description && includeDescription) {
      output += `${baseIndent}// ${description}\n${baseIndent}`
    }
    return output + `${fieldPrefix}${innerType}${comma}`
  }

  // Handle basic types
  let output = ''
  if (description && includeDescription) {
    output += `${baseIndent}// ${description}\n${baseIndent}`
  }
  if (schema instanceof z.ZodString) {
    return output + `${fieldPrefix}string${comma}`
  }
  if (schema instanceof z.ZodBoolean) {
    return output + `${fieldPrefix}boolean${comma}`
  }
  if (schema instanceof z.ZodNumber) {
    return output + `${fieldPrefix}number${comma}`
  }
  if (schema instanceof z.ZodAny) {
    return output + `${fieldPrefix}any${comma}`
  }

  // Fallback for unknown types
  return output + `${fieldPrefix}unknown${comma}`
}

/**
 * Generates a human-readable string representation of a Zod schema,
 * including descriptions defined using `.describe()`.
 *
 * @param schema The Zod schema object (e.g., z.object({...})).
 * @returns A string describing the schema structure and field descriptions.
 */
export function stringifySchema(schema: z.ZodObject<any>): string {
  return stringifyZodType(schema, 0)
}
