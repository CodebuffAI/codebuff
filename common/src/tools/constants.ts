import type { ToolResultPart } from 'ai'

import z from 'zod/v4'

export const toolNameParam = 'cb_tool_name'
export const endsAgentStepParam = 'cb_easp'
export const toolXmlName = 'codebuff_tool_call'
export const startToolTag = `<${toolXmlName}>\n`
export const endToolTag = `\n</${toolXmlName}>`

// List of all available tools
export const toolNames = [
  'add_subgoal',
  'add_message',
  'browser_logs',
  'code_search',
  'create_plan',
  'end_turn',
  'find_files',
  'read_docs',
  'read_files',
  'run_file_change_hooks',
  'run_terminal_command',
  'send_agent_message',
  'set_messages',
  'set_output',
  'spawn_agents',
  'spawn_agents_async',
  'str_replace',
  'think_deeply',
  'update_subgoal',
  'web_search',
  'write_file',
] as const

export type ToolName = (typeof toolNames)[number]

export type ToolParams<T extends ToolName = ToolName> = {
  toolName: T
  endsAgentStep: boolean
  parameters: z.ZodType
}

export type StringToolResultPart = Omit<ToolResultPart, 'type'> & {
  result: string
}
