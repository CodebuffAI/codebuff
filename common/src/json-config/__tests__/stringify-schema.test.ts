// @ts-ignore
import { describe, expect, it } from 'bun:test'
import { z } from 'zod'

import { CodebuffConfigSchema, StartupProcessSchema } from '../constants'
import { stringifySchema } from '../stringify-schema'

describe('stringifySchema', () => {
  it('should correctly stringify StartupProcessSchema', () => {
    const result = stringifySchema(StartupProcessSchema)
    expect(result).toMatchSnapshot()
  })

  it('should correctly stringify CodebuffConfigSchema', () => {
    const result = stringifySchema(CodebuffConfigSchema)
    expect(result).toMatchSnapshot()
  })

  it('should handle a more complex schema', () => {
    const ComplexSchema = z
      .object({
        id: z.string().describe('Unique identifier'),
        count: z.number().int().positive().describe('A positive integer count'),
        isActive: z.boolean().describe('Activity status'),
        tags: z.array(z.string()).optional().describe('Optional list of tags'),
        nested: z
          .object({
            value: z.string(),
            config: z
              .object({
                retries: z.number().default(3).describe('Number of retries'),
              })
              .describe('Nested configuration'),
          })
          .describe('A nested object structure'),
      })
      .describe('A complex test schema')

    const result = stringifySchema(ComplexSchema)
    expect(result).toMatchSnapshot()
  })

  it('should handle an empty object schema', () => {
    const EmptySchema = z.object({}).describe('An empty schema')
    const result = stringifySchema(EmptySchema)
    expect(result).toMatchSnapshot()
  })

  it('should handle schema with only optional fields', () => {
    const OptionalOnlySchema = z
      .object({
        field1: z.string().optional().describe('Optional field 1'),
        field2: z.number().optional().describe('Optional field 2'),
      })
      .describe('Schema with only optional fields')
    const result = stringifySchema(OptionalOnlySchema)
    expect(result).toMatchSnapshot()
  })

  it('should handle schema with default values', () => {
    const DefaultSchema = z
      .object({
        name: z.string().default('anonymous').describe('Name with default'),
        level: z.number().default(1).describe('Level with default'),
        enabled: z.boolean().default(false).describe('Enabled with default'),
      })
      .describe('Schema demonstrating default values')
    const result = stringifySchema(DefaultSchema)
    expect(result).toMatchSnapshot()
  })
})
