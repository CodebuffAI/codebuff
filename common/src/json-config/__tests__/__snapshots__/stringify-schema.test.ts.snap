// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`stringifySchema should correctly stringify StartupProcessSchema 1`] = `
"// Defines a single startup process.
{
  // A user-friendly name for the process. Should be one word and unique.
  "name": string,

  // The actual shell command to execute.
  "command": string,

  // (optional): The working directory from which to run the command.
  "cwd": string | undefined,

  // (optional): Whether this process should be run, default: true
  "enabled": boolean | undefined,

  // (optional): Path to write the process's stdout. If not specified, stderr is not stored.
  "stdoutFile": string | undefined,

  // (optional): Path to write the process's stderr. If not specified, stderr will be put into the stdoutFile.
  "stderrFile": string | undefined
}"
`;

exports[`stringifySchema should correctly stringify CodebuffConfigSchema 1`] = `
"// Defines the overall Codebuff configuration file (e.g., codebuff.json). This schema defines the top-level structure of the configuration. This schema can be found at https://www.codebuff.com/config
{
  // (optional): Does nothing. Put any thing you want here!
  "description": any | undefined,

  // (optional): An array of startup processes.
  "startupProcesses": [

    // Defines a single startup process.
    {
      // A user-friendly name for the process. Should be one word and unique.
      "name": string,

      // The actual shell command to execute.
      "command": string,

      // (optional): The working directory from which to run the command.
      "cwd": string | undefined,

      // (optional): Whether this process should be run, default: true
      "enabled": boolean | undefined,

      // (optional): Path to write the process's stdout. If not specified, stderr is not stored.
      "stdoutFile": string | undefined,

      // (optional): Path to write the process's stderr. If not specified, stderr will be put into the stdoutFile.
      "stderrFile": string | undefined
    }
  ] | undefined,

  // (optional): An array of commands to run on file changes.
  "fileChangeHooks": [

    // Defines a single file change hook.
    {
      // A user-friendly name for the hook. Should be one word and unique.
      "name": string,

      // The actual shell command to execute.
      "command": string,

      // (optional): The working directory from which to run the command.
      "cwd": string | undefined,

      // (optional): Glob pattern to match files.
      "filePattern": string | undefined,

      // (optional): Whether this command should be run, default: true
      "enabled": boolean | undefined
    }
  ] | undefined,

  // (optional): Maximum number of turns agent will take before being forced to end, default: 12
  "maxAgentSteps": number,

  // (optional): Specify default base agent
  "baseAgent": string | undefined,

  // (optional): Specify complete list of spawnable agents for the base agent
  "spawnableAgents": [

  string
  ] | undefined
}"
`;

exports[`stringifySchema should handle a more complex schema 1`] = `
"// A complex test schema
{
  // Unique identifier
  "id": string,

  // A positive integer count
  "count": number,

  // Activity status
  "isActive": boolean,

  // (optional): Optional list of tags
  "tags": [

  string
  ] | undefined,

  // A nested object structure
  "nested": {
  "value": string,

  // Nested configuration
  "config": {
  // (optional): Number of retries, default: 3
  "retries": number
}
}
}"
`;

exports[`stringifySchema should handle an empty object schema 1`] = `
"// An empty schema
{
}"
`;

exports[`stringifySchema should handle schema with only optional fields 1`] = `
"// Schema with only optional fields
{
  // (optional): Optional field 1
  "field1": string | undefined,

  // (optional): Optional field 2
  "field2": number | undefined
}"
`;

exports[`stringifySchema should handle schema with default values 1`] = `
"// Schema demonstrating default values
{
  // (optional): Name with default, default: "anonymous"
  "name": string,

  // (optional): Level with default, default: 1
  "level": number,

  // (optional): Enabled with default, default: false
  "enabled": boolean
}"
`;
